src/toolchain.rs

```diff
@@ -156,23 +161,44 @@ fn init_toolchain_from_dist(toolchain: &str) -> Result<()> {
                    })
 }
 
-fn init_toolchain_from_repo(repo: &str, sha: &str) -> Result<()> {
-    info!("installing toolchain {}#{}", repo, sha);
+fn init_toolchain_from_try(sha: &str) -> Result<()> {
+    info!("installing toolchain try#{}", sha);
 
     fs::create_dir_all(TOOLCHAIN_DIR)?;
-    let dir = &Path::new(TOOLCHAIN_DIR).join(sha);
-    git::shallow_clone_or_pull(repo, dir)?;
-    git::shallow_fetch_sha(repo, dir, sha)?;
-    git::reset_to_sha(dir, sha)?;
+    let dir = Path::new(TOOLCHAIN_DIR).join(sha);
+    use rustup_dist as dist;
+    use rustup_dist::component::Package;
+
+    let prefix = dist::prefix::InstallPrefix::from(dir);
+    let target = dist::component::Components::open(prefix.clone())?;
+    let cfg = dist::temp::Cfg::new(RUSTUP_HOME.into(), RUSTUP_BASE_URL, Box::new(|_| {}));
+    let notifier = |_: dist::notifications::Notification| {};
+    let mut tx = dist::component::Transaction::new(prefix, &cfg, &notifier);
+
+    for &(component, file) in &RUST_CI_COMPONENTS {
+        if target.find(component)?.is_some() {
+            info!("skipping component {}, already installed", component);
+            continue;
+        };
+        info!("installing component {}", component);
+        let url = format!("{}/{}/{}", RUST_CI_TRY_BASE_URL, sha, file);
+        let response = dl::download_limit(&url, 10000)?;
+        if *response.status() != reqwest::StatusCode::Ok {
+            return Err(ErrorKind::Download.into());
+        }
+        tx = dist::component::TarGzPackage::new(response, &cfg)?
+            .install(&target, component, None, tx)?;
+    }
+    tx.commit();
```

Comment:
Heh, this could actually do exactly what it's doing to install rustup to some non-rustup location (instead of calling install.sh), then call 'toolchain link' to turn it into a rustup toolchain.